halo2-ecc/src/bigint/carry_mod.rs

```diff
@@ -56,11 +56,11 @@ pub fn crt<F: BigPrimeField>(
     // Let n' <= quot_max_bits - n(k-1) - 1
     // If quot[i] <= 2^n for i < k - 1 and quot[k-1] <= 2^{n'} then
     // quot < 2^{n(k-1)+1} + 2^{n' + n(k-1)} = (2+2^{n'}) 2^{n(k-1)} < 2^{n'+1} * 2^{n(k-1)} <= 2^{quot_max_bits - n(k-1)} * 2^{n(k-1)}
-    let quot_last_limb_bits = quot_max_bits - n * (k - 1);
-
+    let bits_wo_last_limb: usize = n * (k - 1);
+    // `has_redunant_limb` will be true when native element can be represented in k-1 limbs, but some cases require an extra limb to carry.
+    // This is only the case for BLS12-381, which requires k=5 and n > 102 because of the check above.
+    let has_redunant_limb = quot_max_bits < bits_wo_last_limb;
```

Comment:
It is better to make a general fix, where you determine the number of limbs needed exactly to represent `Fp`: this should be `out_num_limbs = ceil(Fp::NUM_BITS / n)`. Then the loop below should be over `out_assigned.iter().take(out_num_limbs)` and `out_assigned[out_num_limbs..]` needs to be constrained to be zero.

Reply:
got it, pushed a generalized version of the fix